Responsive image aspect ratio variants

jmuzina opened the pull request:
## Done

Adds new `-on-(large|medium|small)` variants of each aspect ratio image container class for more precise control over aspect ratios, depending on the screen size.

Fixes #5274, [WD-13920](https://warthogs.atlassian.net/browse/WD-13920)

## QA

- Open the [responsive-all example](https://vanilla-framework-5308.demos.haus/docs/examples/patterns/image/container/aspect-ratio/responsive-all?theme=light) and verify that the aspect ratio labels above each example there are correct by resizing the window. 
  - Make sure to check at the breakpoints. As long as there are not two classes for the same breakpoint, no two responsive aspect ratio classes should be active.
    - Large: `width >=1036px`
    - Medium: `620px >= width < 1036px`
    - Small: `width < 620px`
- Verify the responsive examples were added to the [combined image example](https://vanilla-framework-5308.demos.haus/docs/examples/patterns/image/combined?theme=light)
- Review [image container with aspect ratio docs](https://vanilla-framework-5308.demos.haus/docs/patterns/images#image-container-with-aspect-ratio)
- Review [4.16.0 release notes](https://vanilla-framework-5308.demos.haus/docs/whats-new)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [ ] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix release (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [ ] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).


[WD-13920]: https://warthogs.atlassian.net/browse/WD-13920?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://vanilla-framework-5308.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/11084)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

jmuzina commented:
I'm not sure what the best way to document the list of aspect ratio classes is here, since there are so many of them (one for each combination of breakpoint and aspect ratio, and another one that applies to all aspect ratios)

I've added a table similar to the [Grid page](https://vanillaframework.io/docs/patterns/grid#common-patterns):

![image](https://github.com/user-attachments/assets/74a52edf-910b-40e8-9e07-c3b4f9e9f74e)

As well as an entry in the class-refs table:
![image](https://github.com/user-attachments/assets/fcb0e3f9-e8e6-4c0d-bb48-886b56d897d4)

lyubomir-popov commented:
> I'm not sure what the best way to document the list of aspect ratio classes is here, since there are so many of them (one for each combination of breakpoint and aspect ratio, and another one that applies to all aspect ratios)

A table is good enough, but I would widen the first column to prevent wrapping. 

Also, might be good to add some explanation as to why we do this - for example, mention that tall aspect ratios can cause a lot of scrolling on mobile, so a change to a less wasteful aspect ratio can be useful. on the other hand, a cinematic format is great when there's sufficient width (at least 9 columns), but for smaller screens, switching to 16x9 or 3x2 might allow more detail to be visible.

jmuzina commented:
> A table is good enough, but I would widen the first column to prevent wrapping.

I've updated the table so that the first column does not wrap on large & medium screens.

<img width="773" alt="Screenshot 2024-08-19 at 3 21 50 PM" src="https://github.com/user-attachments/assets/fe6504da-fedf-4f22-9286-a9b2d9dac0e2">

> Also, might be good to add some explanation as to why we do this - for example, mention that tall aspect ratios can cause a lot of scrolling on mobile, so a change to a less wasteful aspect ratio can be useful. on the other hand, a cinematic format is great when there's sufficient width (at least 9 columns), but for smaller screens, switching to 16x9 or 3x2 might allow more detail to be visible.

I've added an explanation for reasons to use these responsive classes.
<img width="655" alt="Screenshot 2024-08-19 at 3 22 31 PM" src="https://github.com/user-attachments/assets/ccac5621-56da-4b92-b347-758b5010d4a0">

jmuzina commented on scss/_patterns_image.scss:
```diff
@@ -18,13 +18,77 @@
+  'square': 1,
+);
+
+$breakpoints: (
```
Good idea, moved it to `_settings_breakpoints`!

jmuzina commented:
Updated to use inequalities rather than `@media ((min|max)-width)`, as with #5315

jmuzina commented:
Need to refine the bounds between the breakpoints slightly

jmuzina commented:
@pastelcyborg I needed to merge `main` to resolve release notes conflicts - can you re-approve and also review Percy build please?
